### ISSUE TITLE
didLoadWithEvents always getting RNCallKeepDidDisplayIncomingCall

### ISSUE THREAD

**Issue by devWaleed** [react-native-webrtc/react-native-callkeep#330]
# Bug report

- [ ] I've checked the [example](https://github.com/react-native-webrtc/react-native-callkeep/tree/master/example) to reproduce the issue.

- Reproduced on:
 - [ ] iOS

## Description
I have tried to register didLoadWithEvents in index.js and as well as App.js which is my first component. Both are returning RNCallKeepDidDisplayIncomingCall even if i press accept button right when callkit ui appears, i don't get acll accept event name here and if i press accept, i don't get answercall event either

## Versions
    - Callkeep: 4.0
    - React Native: 0.63.3
    - iOS: 14
    - Phone model: iPhone 7 

## Logs

```
Paste here
```

**Comment by devWaleed** [comment]
Just updated my app to latest version of VoipPushNotifications and I am able to receive multiple events of VOIP package correctly but not for CallKeep

**Comment by mikeangelsilva** [comment]
Hi, just wanna ask in ios  , i send a voip notification. Do i need to build the app/archived or is it working during development? 

Im expecting to open/relaunch the app from background/killed?

Thoughts on this please.
Thank you

**Comment by devWaleed** [comment]
@mikeangelsilva Hey, if you're using APN with VOIP type make sure you're using correctly URLs for sending notification i.e for development build you will use development certificate and URLs, and when you upload to Testflight you'll use production certificates and URLs.

**Comment by Stephen-HWJ** [comment]
> Just updated my app to latest version of VoipPushNotifications and I am able to receive multiple events of VOIP package correctly but not for CallKeep
> 
> @manuquentin sir i'll really appreciate your help here. Any comment would be very useful for us.

Hi I am having a similar issue. The VOIP works fine, but Callkeep does not. In my case, even didLoadWithEvents is not called. Did you find any solution to this? Thanks for your help!

**Comment by m4r00p** [comment]
Same here any luck with solving that? it is just reporting: RNCallKeepDidDisplayIncomingCall even if user tap accept or reject.

**Comment by namnm** [comment]
> RNCallKeepDidDisplayIncomingCall even if user tap accept or reject.

Can you put some log to see how many `RNCallKeepDidDisplayIncomingCall` event was triggered? If it's one event for each call, then it should be correct

**Comment by m4r00p** [comment]
Yes, it is a single RNCallKeepDidDisplayIncomingCall for each call which is correct. But incorrect or missing behavior is NOT calling a certain listener answer/end or adding those events in didLoadWithEvents array with js was not ready.

**Comment by Stephen-HWJ** [comment]
> > RNCallKeepDidDisplayIncomingCall even if user tap accept or reject.
> 
> Can you put some log to see how many `RNCallKeepDidDisplayIncomingCall` event was triggered? If it's one event for each call, then it should be correct

Also, I tested that the answerCall in Obj-C somehow did not get triggered. I have no idea what is wrong.

**Comment by namnm** [comment]
Strange that Im using the latest and it likely never happens. Im curious that how did you register the handlers, in hook, or didmount, or directly in the bundle?

**Comment by m4r00p** [comment]
I have them directly in the index.js like this:

`
RNCallKeep.addEventListener('answerCall', onAnswerCallAction)
RNCallKeep.addEventListener('endCall', onEndCallAction)
RNCallKeep.addEventListener('didDisplayIncomingCall', onIncomingCallDisplayed)
`

And also I was debugging Obj-C and answerCall was not called for me. Exactly same as for @Stephen-HWJ

**Comment by m4r00p** [comment]
Any luck with reproducing the issue? Are you testing on iOS 14.3?

**Comment by Stephen-HWJ** [comment]
> Any luck with reproducing the issue? Are you testing on iOS 14.3?

Not really. I am testing on 14.3

**Comment by m4r00p** [comment]
@Stephen-HWJ and you are getting the correct user action event in didLoadWithEvents besides just RNCallKeepDidDisplayIncomingCall? Can you paste or share the code you are using to test?

**Comment by m4r00p** [comment]
@Stephen-HWJ Would you mind posting your setup so we can push the issue forward? It is kind of holding us down and I believe more people may experience this issue.

**Comment by m4r00p** [comment]
BTW. Guys @namnm @devWaleed any luck with solving this issue?

**Comment by Stephen-HWJ** [comment]
> @Stephen-HWJ Would you mind posting your setup so we can push the issue forward? It is kind of holding us down and I believe more people may experience this issue.

Please see the code:
```javascript
    RNCallKeep.addEventListener(
      'didLoadWithEvents',
      async (events) => {    
        const validEvents = compact(events);
        let callDataToAdd = null;
        let callDataToAnswer = null;
        let callDataToReject = null;
        let callDataToInitiateCall = null;
        each(compact(validEvents), event => {
          const { name, data } = event;
          if (name === 'RNCallKeepDidDisplayIncomingCall') {
            callDataToAdd = data;
          }
          if (name === 'RNCallKeepPerformAnswerCallAction') {
            callDataToAnswer = data;
          }
          if (name === 'RNCallKeepPerformEndCallAction') {
            callDataToReject = data;
          }
        });

        const lastEventName = get(last(validEvents), 'name');
        if (lastEventName === 'RNCallKeepDidReceiveStartCallAction') {
          callDataToInitiateCall = get(last(validEvents), 'data');
          // callDataToAnswer = null;
          if (!callDataToReject) {
            callDataToAdd = null;
          }
        }
        if (lastEventName === 'RNCallKeepPerformAnswerCallAction') {
          callDataToAnswer = get(last(validEvents), 'data');
        }

        if (callDataToAdd) {
          didDisplayIncomingCallEvent(callDataToAdd);
          if (callDataToReject) {
            cancelCallHandler(callDataToReject);
          }
          if (callDataToAnswer) {
            answerCallEvent(callDataToAnswer);
          }
        }
        if (callDataToInitiateCall) {
          // this.onStartCall(callDataToAnswer);
        }
      });
```

**Comment by Jerome91410** [comment]
Same issue.
`didLoadWithEvents` only contains `RNCallKeepDidDisplayIncomingCall`.

Checked into Objective C, it seems that the `CXAnswerCallAction` is not called.

It only appears when I receive a push notif and the app is killed.
There is a race somewhere, if i accept the call before the js loaded, then the `AnswerCall` is not reached.

 ```
- Callkeep: 4.0.1
- React Native: 0.61.5
- iOS: 14.4
- Phone model: iPhone 8
```


Reproduction method:

1. be sure that your app is killed
2. send a push voip notif to your app
3. accept the call before js was fully loaded 
4. The js never receive the `RNCallKeepPerformAnswerCallAction`, even in `didLoadWithEvents`

When the issue is reproduced:
- i checked the call's value of `hasConnected` from `callObserver`:
  - hasEnded is equals to `false`
  - hasConnected is equals to `false`
- verified all events received by the module from the beginning , and what i remarked is `sendEventWithNameWrapper` never received the `RNCallKeepPerformAnswerCallAction` event


If i wait that js is fully loaded, then the event is correctly received

**Comment by Jerome91410** [comment]
@devWaleed could you check with the PR if your issue is fixed?
thx

**Comment by namnm** [comment]
@m4r00p Im trying the above PR

**Comment by m4r00p** [comment]
@Jerome91410 @namnm I've just patched this file locally. Yes, the @Jerome91410's fix is addressing this issue perfectly.

Great thanks to all you guys!!!

**Comment by stephanoparaskeva** [comment]
> @Jerome91410 @namnm I've just patched this file locally. Yes, the @Jerome91410's fix is addressing this issue perfectly.
> 
> Great thanks to all you guys!!!

Can you explain how to do this?

**Comment by stephanoparaskeva** [comment]
Our application is blocked by this bug.

**Comment by stephanoparaskeva** [comment]
> @m4r00p Im trying the above PR
> My forked repo: https://github.com/brekekesoftware/react-native-callkeep

Did it work?

**Comment by namnm** [comment]
@stephanoparaskeva In your package.json you have something like this:
```
  "dependencies": {
    ...
    "react-native-callkeep": "^4.0.1",
    ...
  }
```
Then change it to
```
  "dependencies": {
    ...
    "react-native-callkeep": "github:Jerome91410/react-native-callkeep#fix-init",
    ...
  }
```
After that try remove the node_module for sure, `npm install` (or `yarn install`), then rerun `pod install` and rerun the project in your XCode.

**Comment by stephanoparaskeva** [comment]
> @devWaleed could you check with the PR if your issue is fixed?
> thx

This did not fix the issue for me, using this PR:
```js
    "react-native-callkeep": "github:react-native-webrtc/react-native-callkeep#pull/365/head",
```

Sometimes I see `RNCallKeepDidPerfrormAnswerCallAction`

Sometimes I do not.

**Comment by stephanoparaskeva** [comment]
>  "react-native-callkeep": "github:Jerome91410/react-native-callkeep#fix-init",

This did not work either 😔

**Comment by namnm** [comment]
@stephanoparaskeva After some more test, I can confirm your issue, although Im using the fork it still happens.
Not sure how can we fix this?

**Comment by Jerome91410** [comment]
@namnm @stephanoparaskeva 

I do not reproduce anymore the issue with this pr, but checked only on iphone 8 and XR.

Maybe you can provide logs about callkeep and a reproduce method ? 

Don't know if there is any impact but i report with the completion handler defined
```
[RNCallKeep reportNewIncomingCall: uuid
                              handle: phonenumber
                          handleType: @"generic"
                            hasVideo: hasVideo
                 localizedCallerName: fullname
                     supportsHolding: YES
                        supportsDTMF: YES
                    supportsGrouping: NO
                  supportsUngrouping: NO
                         fromPushKit: YES
                             payload: data
               withCompletionHandler: completion];
```

**Comment by namnm** [comment]
@Jerome91410 Just tested on iphone6, still happens. Tried again on iphoneXs but can not reproduce. May be because newer phones have more processing power and load the js bridge faster? Then the code may be still having issue.

**Comment by stephanoparaskeva** [comment]
> @namnm @stephanoparaskeva
> 
> I do not reproduce anymore the issue with this pr, but checked only on iphone 8 and XR.
> 
> Maybe you can provide logs about callkeep and a reproduce method ?
> 
> Don't know if there is any impact but i report with the completion handler defined
> 
> ```
> [RNCallKeep reportNewIncomingCall: uuid
>                               handle: phonenumber
>                           handleType: @"generic"
>                             hasVideo: hasVideo
>                  localizedCallerName: fullname
>                      supportsHolding: YES
>                         supportsDTMF: YES
>                     supportsGrouping: NO
>                   supportsUngrouping: NO
>                          fromPushKit: YES
>                              payload: data
>                withCompletionHandler: completion];
> ```

Have tried with this config on both IPhone 6 and IPhone 12 Pro and it does not work:
```js
0
  [RNCallKeep reportNewIncomingCall: uuid
                             handle: handle
                         handleType: @"generic"
                           hasVideo: YES
                localizedCallerName: callerName
                    supportsHolding: YES
                       supportsDTMF: YES
                   supportsGrouping: YES
                 supportsUngrouping: YES
                        fromPushKit: YES
                            payload: payload.dictionaryPayload
              withCompletionHandler: completion];
  }
```

**Comment by stephanoparaskeva** [comment]
Please can someone help us with this bug, the bug is blocking and there's no information on how to solve it

**Comment by Romick2005** [comment]
Based on @Jerome91410 fork I created my own https://github.com/Romick2005/react-native-callkeep.

What problem I solved:
1) Init CallKeep on native site, not on JS.
2) Pass a function that will called at any call reported event. And make them public via native interface like: RNCallKeepPerformEndCallAction, RNCallKeepDidDisplayIncomingCall, ... etc

AppDelegate:
`-(BOOL) application: (UIApplication *) application didFinishLaunchingWithOptions: (NSDictionary *) launchOptions
{
#ifdef FB_SONARKIT_ENABLED
  InitializeFlipper(application);
#endif

  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate: self launchOptions: launchOptions];


  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge: bridge moduleName: @"{appName}" initialProperties: nil];

  rootView.backgroundColor = [[UIColor alloc] initWithRed: 1.0f green: 1.0f blue: 1.0f alpha: 1];

  [[AVAudioSession sharedInstance] setCategory: AVAudioSessionCategoryPlayback error: nil];

  self.window = [[UIWindow alloc] initWithFrame: [UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];

  [[FBSDKApplicationDelegate sharedInstance] application: application
    didFinishLaunchingWithOptions: launchOptions];

  // Callkeep setup
  // We have to set appName as settings for CallKeep in standardUserDefaults, because we may not call RNCallKeep.setup(options) on JS side
  NSDictionary *settings = [[NSUserDefaults standardUserDefaults] dictionaryForKey:@"RNCallKeepSettings"];

  NSBundle *bundle = [NSBundle mainBundle];
  NSDictionary *info = [bundle infoDictionary];

  settings = @{
    @"appName": [info objectForKey:@"CFBundleDisplayName"]
  };

  RNCallKeep *callKeep = [RNCallKeep allocWithZone: nil];
  [callKeep initCallKitProvider: settings
    // withEventHandler: nil
    withEventHandler: ^(NSString *eventName, id data) {
     
      NSLog (@"[RNCallKeep][setup] event eventName = %@, data = %@", eventName, data);

      if (eventName == RNCallKeepHandleStartCallNotification) {
        NSLog (@"[RNCallKeep][withEventHandler] RNCallKeepHandleStartCallNotification");
      } else if (eventName == RNCallKeepDidReceiveStartCallAction) {
        NSLog (@"[RNCallKeep][withEventHandler] RNCallKeepDidReceiveStartCallAction");
      } else if (eventName == RNCallKeepPerformAnswerCallAction) {
        NSLog (@"[RNCallKeep][withEventHandler] RNCallKeepPerformAnswerCallAction");
      } else if (eventName == RNCallKeepPerformEndCallAction) {

        // Prepare data for http request
        NSString *callUUID = data[@"callUUID"];
        NSString *cancelUrl = "{http://server/call/end}"; // TODO set your server url
        NSDictionary *parameters = @{
          @"callUUID": callUUID,
          @"reasonId": @6
        };

        // Send http delete request to reject call
        NSData *callEndData = [NSJSONSerialization dataWithJSONObject: parameters options: NSJSONWritingPrettyPrinted error: nil];
        NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
        [request setURL: [NSURL URLWithString: cancelUrl]];
        [request setHTTPMethod: @"DELETE"];
        [request addValue: @"application/json" forHTTPHeaderField: @"Content-Type"];
        NSString *postLength = [NSString stringWithFormat: @"%lu", [callEndData length]];
        [request setValue: postLength forHTTPHeaderField: @"Content-Length"];
        [request setHTTPBody: callEndData];
        [[[NSURLSession sharedSession] dataTaskWithRequest: request] resume];
        [callUUIDCallDataMap removeObjectForKey: callUUID];

        NSLog (@"[RNCallKeep][withEventHandler] RNCallKeepPerformEndCallAction");

        [NSThread sleepForTimeInterval:2.0f];
      } else if (eventName == RNCallKeepDidActivateAudioSession) {
        NSLog (@"[RNCallKeep][withEventHandler] RNCallKeepDidActivateAudioSession");
      } else if (eventName == RNCallKeepDidDeactivateAudioSession) {
        NSLog (@"[RNCallKeep][withEventHandler] RNCallKeepDidDeactivateAudioSession");
      } else if (eventName == RNCallKeepDidDisplayIncomingCall) {
        NSString *callUUID = data[@"callUUID"];
        [callUUIDCallDataMap setObject: data forKey: callUUID];

        NSLog (@"[RNCallKeep][withEventHandler] RNCallKeepDidDisplayIncomingCall");
      } else if (eventName == RNCallKeepDidPerformSetMutedCallAction) {
        NSLog (@"[RNCallKeep][withEventHandler] RNCallKeepDidPerformSetMutedCallAction");
      } else if (eventName == RNCallKeepPerformPlayDTMFCallAction) {
        NSLog (@"[RNCallKeep][withEventHandler] RNCallKeepPerformPlayDTMFCallAction");
      } else if (eventName == RNCallKeepDidToggleHoldAction) {
        NSLog (@"[RNCallKeep][withEventHandler] RNCallKeepDidToggleHoldAction");
      } else if (eventName == RNCallKeepProviderReset) {
        NSLog (@"[RNCallKeep][withEventHandler] RNCallKeepProviderReset");
      } else if (eventName == RNCallKeepCheckReachability) {
        NSLog (@"[RNCallKeep][withEventHandler] RNCallKeepCheckReachability");
      } else if (eventName == RNCallKeepDidLoadWithEvents) {
        NSLog (@"[RNCallKeep][withEventHandler] RNCallKeepDidLoadWithEvents");
      } else {
        NSLog (@"[RNCallKeep][withEventHandler] Event %@not found.", eventName);
      }
    }
  ];

  return YES;
}`

@stephanoparaskeva can you please check if my approach work for you?

**Comment by stephanoparaskeva** [comment]
@Romick2005 I tried this exactly as you have it, and installed your version of the repo. deleted node modules, but `didLoadWithEvents` still doesnt have any of the correct events. So gutted

What else can I try?

**Comment by Romick2005** [comment]
> Have tried with on both IPhone 6 and IPhone 12 Pro and it does not work:
> 
> 
> 
> Have tried this with versions:
> 
> ```js
> 
>     "react-native-callkeep": "github:brekekesoftware/react-native-callkeep",
> 
>     "react-native-callkeep": "github:Jerome91410/react-native-callkeep#fix-init",
> 
>     "react-native-voip-push-notification": "^3.1.0",
> 
> ```

You should use my forked repo

**Comment by stephanoparaskeva** [comment]
> > Have tried with on both IPhone 6 and IPhone 12 Pro and it does not work:
> > Have tried this with versions:
> > ```js
> >     "react-native-callkeep": "github:brekekesoftware/react-native-callkeep",
> > 
> >     "react-native-callkeep": "github:Jerome91410/react-native-callkeep#fix-init",
> > 
> >     "react-native-voip-push-notification": "^3.1.0",
> > ```
> 
> You should use my forked repo

I have tried with 
```js
    "react-native-callkeep": "github:Romick2005/react-native-callkeep",

**Comment by Romick2005** [comment]
Are you able to run the app in xcode? Have you try to set a breakpoints? What events do you have when js bundle starts? I can't help you as I do not know what problem so you have. For me it just works.

**Comment by stephanoparaskeva** [comment]
> Are you able to run the app in xcode? Have you try to set a breakpoints? What events do you have when js bundle starts? I can't help you as I do not know what problem so you have. For me it just works.

I'm not the best when working with XCODE, I'm more experienced with JS.

Is there any chance I can pay you to have a Teams Meeting where we discuss this and debug the issue together?

My email is stephanokparaskeva@gmail.com if you are available at all this week, let me know and we can arrange a price and a time.

Thank you!

**Comment by Romick2005** [comment]
> @Jerome91410 Just tested on iphone6, still happens. Tried again on iphoneXs but can not reproduce. May be because newer phones have more processing power and load the js bridge faster? Then the code may be still having issue.

Tested on iPhone 5s and iPhone XS cannot reproduce any issues. 
PS on js side I do just do a loop for all received events in didLoadWithEvents. As the last event for me was not AswerCall but something like StartAudioSession. So you need carefully inspect all provided events with data.

**Comment by Jerome91410** [comment]
@stephanoparaskeva To provide logs:
Launch Xcode, install the app on your real device, then `CMD + 2`.
It will open a window where you have to select your device in the `Devices` list and click on `Open Console`

A console will appear, you have to filter on `RNCallKeep` and `Start` the logger.

Here after what a log looks when receiving an incoming call 
<img width="592" alt="Capture d’écran 2021-03-16 à 16 29 41" src="https://user-images.githubusercontent.com/44202034/111335637-d69eed00-8674-11eb-9241-2cafa2d78802.png">

**Comment by namnm** [comment]
@Jerome91410 Your log looks interesting to me. I have an issue with our app is that the call has no audio. At first I didnt think it was related to this package.

Our QA tested the app and found that the call has no audio sometimes. To reproduce it, we need to follow exactly some steps like:
- Kill app
- Open app and login (user A)
- Switch A to run in background mode, open another application like a web browser
- Wait for **about 10s** since A has switched to run in background
- User B still sees A is online, B start a call to A
- VoIP PN comes to A, and the phone ring
- A let the PN ringing in **about 5s** then select Answer to pickup
- App is loaded and displayed on A's side. It shows call is connected, but there is no voice on both sides

Timing is important to reproduce the issue, but it may due to some logic of our app.
Any idea that the issue can be related to this library?

**Comment by stephanoparaskeva** [comment]
This is the only fork that fixes some bugs in my app:
```js
    "react-native-callkeep": "github:Romick2005/react-native-callkeep",
```

DidLoadWithEvents still doesnt have answer call event but a lot of my bugs (like callkeep UI never ending) were all fixed from this fork.

I have tried all the other forks and there were not any fixes for didLoadWithEvents or my bugs.
So I'm just going to use this fork from @Romick2005 

Would encourage others to try what he has posted here:
https://github.com/react-native-webrtc/react-native-callkeep/issues/330#issuecomment-799525290

**Comment by namnm** [comment]
@Jerome91410 What do you think?

Also can you make enable Slack invitation link and post the link here so everyone can join?

**Comment by stephanoparaskeva** [comment]
> @Jerome91410 What do you think?
> 
> 
> 
> Also can you enable Slack invitation link and post the link here so everyone can join?

I think some sort of discord group or slack group for us all to discuss would be great, these issues are really tough and would be great to start a small community. I'm sure it'd speed up all of our lives

**Comment by namnm** [comment]
https://join.slack.com/t/callkeep/shared_invite/zt-nxb6j2yf-VUfINh~Y~hPZz0lq9GIuBQ

**Comment by denniswon** [comment]
@Romick2005 I tried your fork (https://github.com/react-native-webrtc/react-native-callkeep/issues/330#issuecomment-799525290) and I do get the native side event being triggered. How do I emit this event to the JS side to handle these events?

**Comment by Romick2005** [comment]
All events will be triggered on js side when it is setup js bridge. You can use didloadwithevents event.

**Comment by denniswon** [comment]
I did that. Here is my code:

```
RNCallKeep.addEventListener(
      'didLoadWithEvents',
      this.onDidLoadWithNativeEvents
    )

onDidLoadWithNativeEvents = (events) => {
    const validEvents = compact(events)

    let callDataToAdd = null
    let callDataToAnswer = null
    let callDataToReject = null
    let callDataToInitiateCall = null

    logger.info('onDidLoadWithNativeEvents', events)

    each(compact(validEvents), event => {
      const { name, data } = event
      if (name === RNCallKeepDidDisplayIncomingCall) {
        callDataToAdd = data
        callDataToAnswer = null
        callDataToReject = null
      }
      if (name === RNCallKeepPerformAnswerCallAction) {
        console.log('@@@@@@@')
        callDataToReject = null
        callDataToAnswer = data
      }
      if (name === RNCallKeepPerformEndCallAction) {
        callDataToAnswer = null
        callDataToReject = data
      }
    })

    const lastEventName = get(last(validEvents), 'name')
    if (lastEventName === RNCallKeepDidReceiveStartCallAction) {
      callDataToInitiateCall = get(last(validEvents), 'data')
      callDataToAnswer = null
      if (!callDataToReject) {
        callDataToAdd = null
      }
    }

    if (callDataToAdd) {
      this.onDisplayIncomingCall(callDataToAdd)
      if (callDataToReject) {
        this.onEndCall(callDataToReject)
      }
      if (callDataToAnswer) {
        this.onAnswerCall(callDataToAnswer)
      }
    }
    if (callDataToInitiateCall) {
      this.onStartCall(callDataToAnswer)
    }
  }
```

But I am not getting answer/reject call action from rncallkeep. Any idea why?
Thank you for your time. @Romick2005

**Comment by Romick2005** [comment]
I would not call RNCallKeep.setup from JS, as it is already done as soon as possible on native part in appDelegate. There are some issues with creating CallKeep singleton so it could be that you have two instances.

**Comment by beqramo** [comment]
@Romick2005 I get this errors when I tried to paste your codes:

![image](https://user-images.githubusercontent.com/40486471/127220740-587cfc80-ca3a-4ea4-b05e-83029d24d6b6.png)
if there is my mistake, sorry for that I'm not really good at Objective C :D

**Comment by Romick2005** [comment]
So first error is that you do not use Facebook lib to login to the app, so you can remove that line. Other two is because I used my own fork as original one contains all functionality incapsulated and there are no chance to get events on native part. So I created my own fork for this purpose.

**Comment by beqramo** [comment]
Okay I will use yours but I have one question in general: why your method isn't merged to this main repo? 
What other solutions exist for this, as I understand there is no solution for handling reject case when app is killed

**Comment by Romick2005** [comment]
- That was what I also asked and also created a pull request. But all in one people say that your js should start the app so there could be WebSocket connection established, but no UI part. I have test this and if there are not enough time for the app to start from the killed state and user press reject call then there is no way to inform other party. That is especially noticeable for my old iPhone 5s, but for newest XS app starts very quickly. So I decided to go with my own forked version 
"react-native-callkeep": "git+https://github.com/Romick2005/react-native-callkeep.git".

**Comment by beqramo** [comment]
Anyway I really appreciate your work and help here. I hope I will handle everything from this point. 

Thank you and have a nice day!

**Comment by Romick2005** [comment]
There are also Slack group so you can read it [here](https://callkeep.slack.com/archives/C01SJ8D25FB/p1616673370008500) also.

**Comment by beqramo** [comment]
Thanks but I'm unable to sign in, it says that my email doesn't exist in that workspace and if you can please add me: bekakokhodze98@gmail.com

**Comment by rajasekhar911** [comment]
@Romick2005 if possible, please add my email also to the slack group, thank you. rajasekhar911@gmail.com

**Comment by samducker** [comment]
@Romick2005 could you add my email to the slack group please sam@callinganyone.com

**Comment by saikiran402** [comment]
Could you please add me to that group sai.kiran3500@gmail.com

**Comment by namnm** [comment]
https://join.slack.com/t/callkeep/shared_invite/zt-ukweahhw-ygI4nbf6b5SsjfDu7nkDvg

**Comment by gabrielelana** [comment]
@namnm found another solution see my last comment to https://github.com/react-native-webrtc/react-native-callkeep/issues/162

**Comment by yunusaydinonline** [comment]
You can setup RNCallKeep natively in your AppDelegate file as:

```
[RNCallKeep setup:@{
    @"appName": @"Awesome App",
    @"maximumCallGroups": @3,
    @"maximumCallsPerCallGroup": @1,
    @"supportsVideo": @NO,
}];
```

Then you start getting other events before load.
I'm not sure if it works in RNCallKeep v4.0 though.

https://github.com/react-native-webrtc/react-native-callkeep#Usage
